cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll

```diff
@@ -58,6 +59,78 @@ class MatchValue extends AbstractValue, TMatchValue {
   override string toString() { result = this.getCase().toString() }
 }
 
+/**
+ * A value number such that at least one of the instructions is
+ * a `CompareInstruction`.
```

Comment:
I'd like to clarify what "at least one of the instructions" means?  Does that mean the comparison could be a child of the `ValueNumber` in question rather than the root, e.g. `(a == b) || c` is a `CompareValueNumber`?

Reply:
No no. Remember that a value number is basically a collection of expressions that all evaluate to the same value - and you can get the members of this collection by calling `.getAnInstruction()` (or `getAnExpr()` if you're using the AST wrapper).

An `CompareValueNumber` is a value number where one of its members is a `CompareInstruction`. For example:
```
bool b = n == 1;
if(b) {
  // ...
}
```
here, there's a value number that represents the set of values {`n == 1`, `b`}, and this value number is a `CompareValueNumber` because one of the elements is a comparison.

Another example is:
```cpp
bool b = n == 1;
const int one = 1;
bool b2 = n == one;
if(b) {
  // ...
}
```
here, the value number that represents the collection of expressions {`n == 1`, `n == one`, `b`} is also a `CompareValueNumber` because one of its elements is a comparison. In fact, _two_ of its members are comparisons.

Reply:
Ah I see.  Yes, this looks like exactly what we want here.